tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -126,6 +127,320 @@ def type_signature(self):
+    zero, plus = tuple(await
+                       asyncio.gather(*[
+                           executor_utils.embed_tf_scalar_constant(
+                               self.executor, arg.type_signature.member, 0),
+                           executor_utils.embed_tf_binary_operator(
+                               self.executor, arg.type_signature.member, tf.add)
+                       ]))
+    return await self.executor._compute_intrinsic_federated_reduce(
+        FederatingExecutorValue(
+            anonymous_tuple.AnonymousTuple([
+                (None, arg.internal_representation),
+                (None, zero.internal_representation),
+                (None, plus.internal_representation)
+            ]),
+            computation_types.NamedTupleType(
+                (arg.type_signature, zero.type_signature, plus.type_signature)))
+    )
+
+  async def federated_mean(self, arg):
+    arg_sum = await self.executor._compute_intrinsic_federated_sum(arg)
+    member_type = arg_sum.type_signature.member
+    count = float(len(arg.internal_representation))
+    if count < 1.0:
+      raise RuntimeError('Cannot compute a federated mean over an empty group.')
+    child = self.executor._target_executors[placement_literals.SERVER][0]
+    factor, multiply = tuple(await asyncio.gather(*[
+        executor_utils.embed_tf_scalar_constant(child, member_type,
+                                                float(1.0 / count)),
+        executor_utils.embed_tf_binary_operator(child, member_type, tf.multiply)
+    ]))
+    multiply_arg = await child.create_tuple(
+        anonymous_tuple.AnonymousTuple([(None,
+                                         arg_sum.internal_representation[0]),
+                                        (None, factor)]))
+    result = await child.create_call(multiply, multiply_arg)
+    return FederatingExecutorValue([result], arg_sum.type_signature)
+  
+  async def federated_weighted_mean(self, arg):
+    return await executor_utils.compute_federated_weighted_mean(
+        self.executor, arg)
+
+  async def federated_collect(self, arg):
+    py_typecheck.check_type(arg.type_signature, computation_types.FederatedType)
+    type_utils.check_federated_type(
+        arg.type_signature, placement=placement_literals.CLIENTS)
+    val = arg.internal_representation
+    py_typecheck.check_type(val, list)
+    member_type = arg.type_signature.member
+    child = self.executor._target_executors[placement_literals.SERVER][0]
+    collected_items = await child.create_value(
+        await asyncio.gather(*[v.compute() for v in val]),
+        computation_types.SequenceType(member_type))
+    return FederatingExecutorValue(
+        [collected_items],
+        computation_types.FederatedType(
+            computation_types.SequenceType(member_type),
+            placement_literals.SERVER))
+
+  async def federated_secure_sum(self, arg):
+    raise NotImplementedError('The secure sum intrinsic is not implemented.')
```

Comment:
these could be part of a base strategy

Reply:
Yeah, true. I didn't want to necessarily add these because it might be useful for people to check e.g. `hasattr('federated_secure_sum', intrinsic_strategy)`, but perhaps not?